rust/ccommon_rs/src/log/shim.rs

```diff
@@ -0,0 +1,226 @@
+struct CCLog;
+
+unsafe impl Send for CCLog {}
+
+unsafe impl Sync for CCLog {}
+
+fn level_to_int(level: Level) -> c_int {
+    use cc_binding::*;
+
+    let level = match level {
+        Level::Trace => LOG_VERB,
+        Level::Debug => LOG_DEBUG,
+        Level::Info => LOG_INFO,
+        Level::Warn => LOG_WARN,
+        Level::Error => LOG_ERROR,
+    };
+
+    level as c_int
+}
+
+impl CCLog {
+    fn enabled_ptr(metadata: &Metadata, ptr: *const debug_logger) -> bool {
+        if ptr.is_null() {
+            return false;
+        }
+
+        unsafe {
+            // Note: Assumes the level doesn't change concurrently
+            (*ptr).level >= level_to_int(metadata.level())
+        }
+    }
+}
+
+impl Log for CCLog {
+    #[inline]
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        Self::enabled_ptr(metadata, ccommon_dlog().load(Ordering::Relaxed))
+    }
+
+    fn log(&self, record: &Record) {
+        let ptr = ccommon_dlog().load(Ordering::Relaxed);
+
+        if !Self::enabled_ptr(record.metadata(), ptr) {
+            return;
+        }
+
+        // +1 bytes are trailing nul bytes
+        let mut buffer = [0; LOG_MAX_LEN as usize + 1];
+        let mut filename = [0; FILENAME_MAX_LEN + 1];
+
+        let _ = write!(&mut buffer[0..LOG_MAX_LEN as usize], "{}", record.args());
+
+        // Make a best-effort attempt to provide a meainingful
+        // location that is consistent with how ccommon does
+        // log messages.
+        let filestr = record
+            .file()
+            .or_else(|| record.module_path())
+            .unwrap_or(record.target());
+        let _ = write!(&mut filename[0..FILENAME_MAX_LEN], "{}", filestr);
+
+        unsafe {
+            _log(
+                ptr,
+                (&filename).as_ptr() as *const c_char,
+                record.line().map(|x| x as c_int).unwrap_or(-1),
+                level_to_int(record.level()),
+                "%s".as_ptr() as *const c_char,
+                (&buffer).as_ptr() as *const c_char,
+            )
+        }
+    }
+
+    fn flush(&self) {
+        unsafe {
+            // The argument here is unused and ccommon uses it for
+            // compatibility with other modules such as timer_wheel
+            debug_log_flush(std::ptr::null_mut());
+        }
+    }
+}
+
+// The logger has no state so a global zero-sized instance
+// works just fine.
+static LOGGER: CCLog = CCLog;
+
+/// Initialize the logger. It will automatically use the
+/// ccommon debug logger once that is set up.
+pub fn init() -> Result<(), SetLoggerError> {
+    use log::LevelFilter;
+
+    // TODO: dynamically set this based on dlog?
+    log::set_max_level(LevelFilter::Trace);
+
+    log::set_logger(&LOGGER)
+}
+
+/// Set the global panic handler to one which logs the
+/// panic at `LOG_CRIT` level in addition to printing
+/// the panic info to stderr.
+///
+/// This is meant to be
+pub fn set_panic_handler() {
+    use backtrace::Backtrace;
```

Comment:
Will the backtrace dependency be compatible w/ MUSL builds? Can we add `x86_64-unknown-linux-musl` target to the CI to make sure? I've had problems w/ backtrace + musl historically, but looks like it might be fixed now?

Reply:
I've redone this to use the standard lib's backtrace printing. That should (hopefully) avoid any issues around musl and the backtrace crate